### PR TITLE
Handle killing of build process properly

### PR DESCRIPTION
**Pull request by alexreg**
Fixes issue #2484. I've tested on macOS Big Sur, of course, and it's working nicely. I've also slightly modified the Windows method. On all platforms now, it should try a graceful termination before a hard kill (after a one second timeout).

I also took the liberty to rename the "Kill" commmand, but obviously you can choose to not use that commit. See the commit log for the reason, however.

CC @jlelong

### CODE REVIEW

**Comment by jlelong** [comment]
Unfortunately, the PR does not work on Catalina because it only kills `latexmk` but not the `pdflatex` processes. The [documentation of the kill method](https://nodejs.org/api/child_process.html#child_process_subprocess_kill_signal) says

> On Linux, child processes of child processes will not be terminated when attempting to kill their parent. This is likely to happen when running a new process in a `shell` or with the use of the shell option of `ChildProcess`

This was already discussed in #1680. Hence, the use of `pkill -P`.

**Comment by jlelong** [comment]
Btw, I am not willing to rename `kill` into `stop` for several reasons

1. This method indeed calls the `kill` method from `child_process`
2. Users will keep looking for a `kill` method, which they will not find

**Comment by alexreg** [comment]
Alright, I will look into that tomorrow, but it should be simple enough to resolve.

Regarding the renaming:
1. An implementation detail should not matter (affect naming). The standard terminology in build processes is "Stop" or occasionally "Cancel".
2. You could always keep that command around for a short transitionary period and have it either alias the "Stop" command or simply display an informational message about the renaming.
Indeed, I spent a while looking for a command to stop the build before realising it was weirdly called "Kill". It would seem sensible to adopt the terminology most consistent across build processes, but that's all I'll say on the matter.

**Comment by alexreg** [comment]
@jlelong See my new push. This solves the issue, I believe. The real issue at the end of the day was that one can have more than one `pkill` installed (I had one in `/usr/local/bin` that was getting used). This is a different version of pkill, that seems to be problematic in that it doesn't like being run from within VS Code (maybe within the same process group?). So this explicitly specifies the `PATH` env var to get around that. I also do `spawn` instead of `exec` to a) prevent a shell possibly interfering with environment variables, b) saving time forking a shell process, even though it's not too much. Apart from that, the changes generally make the process of stopping/killing the build more robust, and like I said before, attempts a graceful termination before a hard kill.

You may want to double check that everything still works on Windows with regards to specifying the PATH variable correctly.

File: src/components/builder.ts [hunk 1]
```diff
@@ -48,24 +48,68 @@ export class Builder {
...
+     * Stop the current build process.
      */
-    kill() {
-        const proc = this.currentProcess
-        if (proc) {
-            const pid = proc.pid
+    async stop() {
+        if (this.currentProcess) {
+            const envVarsPosix: ProcessEnv = {PATH: '/bin:/usr/bin'}
+            const envVarsWindows: ProcessEnv = {PATH: '%SystemRoot%\\system32;%SystemRoot%;%SystemRoot%\\System32\\Wbem'}
```
Comment (jlelong) [inline src/components/builder.ts:56]: I am against changing the `PATH` on Windows. I do not know how canonical they are on Windows and there is hardly no chance that there is an other `pkill` executable available in the `PATH`. Please, revert these changes.

File: src/components/builder.ts [hunk 2]
```diff
@@ -48,24 +48,68 @@ export class Builder {
...
+
+            const currentProcessExit = new Promise(listener => this.currentProcess?.on('exit', listener))
+            const pid = this.currentProcess.pid
+
+            // Try to terminate the build process gracefully first.
             if (process.platform === 'linux' || process.platform === 'darwin') {
-                cp.execSync(`pkill -P ${pid}`)
+                try {
+                    // Terminate build process and all its child processes.
+                    await this.spawnAndWait('pkill', ['-TERM', '-P', pid.toString()], {env: envVarsPosix})
```
Comment (jlelong) [inline src/components/builder.ts:65]: Use [`spawnSync`](https://nodejs.org/api/child_process.html#child_process_child_process_spawnsync_command_args_options) instead of defining your own `spawnAndWait`.

File: src/components/builder.ts [hunk 3]
```diff
@@ -48,24 +48,68 @@ export class Builder {
...
+                    await this.spawnAndWait('kill', ['-TERM', pid.toString()], {env: envVarsPosix})
```
Comment (jlelong) [inline src/components/builder.ts:66]: This is exactly what `this.currentProcess.kill()` does. Why not use it?

**Review by jlelong — CHANGES_REQUESTED** [review]

**Comment by jlelong** [comment]
As the PR has become quite complicated, I suggest to do one thing at a time and postpone the renaming to an other PR once  #2484 is fixed. 

Concerning the core of this PR,
- I am against changing the `PATH` on Windows. I do not know how canonical they are on Windows and there is hardly no chance that there is an other `pkill` executable available in the `PATH`. Please, revert these changes.
- The default signal used by the `kill` & co functions is `SIGTERM` so I do not think your approach is more graceful.

My general feeling is that the real issue is more on the [proctools](http://proctools.sourceforge.net/)  you have in `/usr/local/bin`. I am not sure how far the extension should go to support that.

File: src/components/builder.ts (hunk 3 above)
Comment (alexreg) [reply src/components/builder.ts:66]: The second line, you mean? It is more consistent with the other `pkill` and `taskkill` commands (in that it runs a program), but I do see your point. I can switch to using `kill` if you stll prefer.

**Comment by alexreg** [comment]
> * I am against changing the PATH on Windows. I do not know how canonical they are on Windows and there is hardly no chance that there is an other pkill executable available in the PATH. Please, revert these changes.

100% canonical, since they rely on %systemroot%. You can check this yourself in the Windows documentation.

> * The default signal used by the kill & co functions is SIGTERM so I do not think your approach is more graceful.

I should be more specific: it is more *robust* in that it tries SIGTERM and then SIGKILL if the former fails. It is more *graceful* in that you were previously automatically doing a hard kill (`/f`) on Windows, which is obviously a bad idea. Furthermore, just using synchronous functions/methods like `execSync` and `spawnSync` is not great, as it can cause VS Code to operate less smoothly (I experienced this during testing).

* I am not 100% sure that pkill is always in /bin or /usr/bin on Unix. Could not it be in /sbin or /usr/sbin?

I am almost positive they are always there from experience, but it's no big deal to add those two to the PATH too. What you don't want in the path is `/usr/local/bin` and `/opt/local/bin` and other stuff.

* My general feeling is that the real issue is more on the proctools you have in /usr/local/bin. Note that proctools has not been updated since 2013. Why not just remove that package from your homebrew installation? I am not sure how far the extension should go to support that pkill is not the standard one.

Telling people what to do with the other software they have installed (and what they put in their PATH) is a very bad idea. Not only will people resent being told, but moreover it is a well-known sound principle of software engineering to avoid conflicts with other installed software and libraries wherever possible. There could be many other scenarios of a `kill` or `pkill` program in the user's PATH that operates differently to the standard Posix one.

**Comment by tamuratak** [comment]
I am also against merging this PR. Please uninstall `proctools`.

**Comment by alexreg** [comment]
@tamuratak You have a real habit of insolence, you know. Such ingratitude. Nor do you understand basic software engineering principles, it seems. Who is in charge of this project? I would like to know that.
